File: app/src/main/java/info/nightscout/androidaps/plugins/PumpDanaR/comm/MsgStatusTempBasal.java
package info.nightscout.androidaps.plugins.PumpDanaR.comm;

import android.support.annotation.NonNull;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Date;

import info.nightscout.androidaps.Config;
import info.nightscout.androidaps.db.Source;
import info.nightscout.androidaps.db.TemporaryBasal;
import info.nightscout.androidaps.plugins.PumpDanaR.DanaRPump;
import info.nightscout.androidaps.plugins.Treatments.TreatmentsPlugin;

public class MsgStatusTempBasal extends MessageBase {
    private static Logger log = LoggerFactory.getLogger(MsgStatusTempBasal.class);

    public MsgStatusTempBasal() {
        SetCommand(0x0205);
    }

    public void handleMessage(byte[] bytes) {
        boolean isTempBasalInProgress = (intFromBuff(bytes, 0, 1) & 0x01) == 0x01;
        boolean isAPSTempBasalInProgress = (intFromBuff(bytes, 0, 1) & 0x02) == 0x02;
        int tempBasalPercent = intFromBuff(bytes, 1, 1);
        if (tempBasalPercent > 200) tempBasalPercent = (tempBasalPercent - 200) * 10;
        int tempBasalTotalSec;
        if (intFromBuff(bytes, 2, 1) == 150) tempBasalTotalSec = 15 * 60;
        else if (intFromBuff(bytes, 2, 1) == 160) tempBasalTotalSec = 30 * 60;
        else tempBasalTotalSec = intFromBuff(bytes, 2, 1) * 60 * 60;
        int tempBasalRunningSeconds = intFromBuff(bytes, 3, 3);
        int tempBasalRemainingMin = (tempBasalTotalSec - tempBasalRunningSeconds) / 60;
        Date tempBasalStart = isTempBasalInProgress ? getDateFromTempBasalSecAgo(tempBasalRunningSeconds) : new Date(0);

        DanaRPump pump = DanaRPump.getInstance();
        pump.isTempBasalInProgress = isTempBasalInProgress;
        pump.tempBasalPercent = tempBasalPercent;
        pump.tempBasalRemainingMin = tempBasalRemainingMin;
        pump.tempBasalTotalSec = tempBasalTotalSec;
        pump.tempBasalStart = tempBasalStart;

        updateTempBasalInDB();

        if (Config.logDanaMessageDetail) {
            log.debug("Is temp basal running: " + isTempBasalInProgress);
            log.debug("Is APS temp basal running: " + isAPSTempBasalInProgress);
            log.debug("Current temp basal percent: " + tempBasalPercent);
            log.debug("Current temp basal remaining min: " + tempBasalRemainingMin);
            log.debug("Current temp basal total sec: " + tempBasalTotalSec);
            log.debug("Current temp basal start: " + tempBasalStart);
        }
    }

    @NonNull
    private Date getDateFromTempBasalSecAgo(int tempBasalAgoSecs) {
        return new Date((long) (Math.ceil(System.currentTimeMillis() / 1000d) - tempBasalAgoSecs) * 1000);
    }

    public static void updateTempBasalInDB() {
        DanaRPump danaRPump = DanaRPump.getInstance();
        long now = System.currentTimeMillis();

        if (TreatmentsPlugin.getPlugin().isInHistoryRealTempBasalInProgress()) {
            TemporaryBasal tempBasal = TreatmentsPlugin.getPlugin().getRealTempBasalFromHistory(System.currentTimeMillis());
            if (danaRPump.isTempBasalInProgress) {
                if (tempBasal.percentRate != danaRPump.tempBasalPercent) {
                    // Close current temp basal
                    TemporaryBasal tempStop = new TemporaryBasal().date(danaRPump.tempBasalStart.getTime() - 1000).source(Source.USER);
                    TreatmentsPlugin.getPlugin().addToHistoryTempBasal(tempStop);
                    // Create new
                    TemporaryBasal newTempBasal = new TemporaryBasal()
                            .date(danaRPump.tempBasalStart.getTime())
                            .percent(danaRPump.tempBasalPercent)
                            .duration(danaRPump.tempBasalTotalSec / 60)
                            .source(Source.USER);
                    TreatmentsPlugin.getPlugin().addToHistoryTempBasal(newTempBasal);
                }
            } else {
                // Close current temp basal
                TemporaryBasal tempStop = new TemporaryBasal().date(now).source(Source.USER);
                TreatmentsPlugin.getPlugin().addToHistoryTempBasal(tempStop);
            }
        } else {
            if (danaRPump.isTempBasalInProgress) {
                // Create new
                TemporaryBasal newTempBasal = new TemporaryBasal()
                        .date(danaRPump.tempBasalStart.getTime())
                        .percent(danaRPump.tempBasalPercent)
                        .duration(danaRPump.tempBasalTotalSec / 60)
                        .source(Source.USER);
                TreatmentsPlugin.getPlugin().addToHistoryTempBasal(newTempBasal);
            }
        }
    }
}

File: danar/src/main/java/info/nightscout/androidaps/danar/comm/MsgStatusTempBasal.kt
package info.nightscout.androidaps.danar.comm

import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.db.Source
import info.nightscout.androidaps.db.TemporaryBasal
import info.nightscout.androidaps.logging.LTag
import kotlin.math.ceil

class MsgStatusTempBasal(
    injector: HasAndroidInjector
) : MessageBase(injector) {

    init {
        SetCommand(0x0205)
        aapsLogger.debug(LTag.PUMPCOMM, "New message")
    }

    override fun handleMessage(bytes: ByteArray) {
        val isTempBasalInProgress = intFromBuff(bytes, 0, 1) and 0x01 == 0x01
        val isAPSTempBasalInProgress = intFromBuff(bytes, 0, 1) and 0x02 == 0x02
        var tempBasalPercent = intFromBuff(bytes, 1, 1)
        if (tempBasalPercent > 200) tempBasalPercent = (tempBasalPercent - 200) * 10
        val tempBasalTotalSec: Int = if (intFromBuff(bytes, 2, 1) == 150) 15 * 60 else if (intFromBuff(bytes, 2, 1) == 160) 30 * 60 else intFromBuff(bytes, 2, 1) * 60 * 60
        val tempBasalRunningSeconds = intFromBuff(bytes, 3, 3)
        val tempBasalRemainingMin = (tempBasalTotalSec - tempBasalRunningSeconds) / 60
        val tempBasalStart = if (isTempBasalInProgress) getDateFromTempBasalSecAgo(tempBasalRunningSeconds) else 0
        danaPump.isTempBasalInProgress = isTempBasalInProgress
        danaPump.tempBasalPercent = tempBasalPercent
        danaPump.tempBasalRemainingMin = tempBasalRemainingMin
        danaPump.tempBasalTotalSec = tempBasalTotalSec
        danaPump.tempBasalStart = tempBasalStart
        updateTempBasalInDB()
        aapsLogger.debug(LTag.PUMPCOMM, "Is temp basal running: $isTempBasalInProgress")
        aapsLogger.debug(LTag.PUMPCOMM, "Is APS temp basal running: $isAPSTempBasalInProgress")
        aapsLogger.debug(LTag.PUMPCOMM, "Current temp basal percent: $tempBasalPercent")
        aapsLogger.debug(LTag.PUMPCOMM, "Current temp basal remaining min: $tempBasalRemainingMin")
        aapsLogger.debug(LTag.PUMPCOMM, "Current temp basal total sec: $tempBasalTotalSec")
        aapsLogger.debug(LTag.PUMPCOMM, "Current temp basal start: $tempBasalStart")
    }

    private fun getDateFromTempBasalSecAgo(tempBasalAgoSecs: Int): Long {
        return (ceil(System.currentTimeMillis() / 1000.0) - tempBasalAgoSecs).toLong() * 1000
    }

    private fun updateTempBasalInDB() {
        val now = System.currentTimeMillis()
        if (activePlugin.activeTreatments.isInHistoryRealTempBasalInProgress) {
            val tempBasal = activePlugin.activeTreatments.getRealTempBasalFromHistory(System.currentTimeMillis())
            if (danaPump.isTempBasalInProgress) {
                if (tempBasal.percentRate != danaPump.tempBasalPercent) { // Close current temp basal
                    val tempStop = TemporaryBasal(injector).date(danaPump.tempBasalStart - 1000).source(Source.USER)
                    activePlugin.activeTreatments.addToHistoryTempBasal(tempStop)
                    // Create new
                    val newTempBasal = TemporaryBasal(injector)
                        .date(danaPump.tempBasalStart)
                        .percent(danaPump.tempBasalPercent)
                        .duration(danaPump.tempBasalTotalSec / 60)
                        .source(Source.USER)
                    activePlugin.activeTreatments.addToHistoryTempBasal(newTempBasal)
                }
            } else { // Close current temp basal
                val tempStop = TemporaryBasal(injector).date(now).source(Source.USER)
                activePlugin.activeTreatments.addToHistoryTempBasal(tempStop)
            }
        } else {
            if (danaPump.isTempBasalInProgress) { // Create new
                val newTempBasal = TemporaryBasal(injector)
                    .date(danaPump.tempBasalStart)
                    .percent(danaPump.tempBasalPercent)
                    .duration(danaPump.tempBasalTotalSec / 60)
                    .source(Source.USER)
                activePlugin.activeTreatments.addToHistoryTempBasal(newTempBasal)
            }
        }
    }
}